### PR TITLE
New snli datasetreader

### PR DESCRIPTION
**Pull request by leo-liuzy**
Add an option to collapse labels "neutral" and "contradiction" into "non-entailment"
This is a common practice in research.

### CODE REVIEW

File: allennlp_models/lm/models/next_token_lm.py [hunk 1]
```diff
@@ -117,7 +117,13 @@ def forward(  # type: ignore
...
         if target_ids is not None:
             batch_size, vocab_size = target_logits.size()
-            targets = util.get_token_ids_from_text_field_tensors(target_ids).view(batch_size)
+            tmp = util.get_token_ids_from_text_field_tensors(target_ids)
+            # In some scenarios, target_ids might be a topk list of token ids (e.g. sorted by probabilities).
+            # Therefore, we need to make sure only one token per batch
+            # Assume: first token in each batch is the most desirable one (e.g. highest probability)
+            tmp = tmp[:, 0] if len(tmp.shape) == 2 else tmp
+            assert len(tmp.shape) <= 2
+            targets = tmp.view(batch_size)
```
Comment (epwalsh) [inline allennlp_models/lm/models/next_token_lm.py:126]: These changes are in https://github.com/allenai/allennlp-models/pull/213/files, right? If so, can you remove them from this PR?

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [v2.1.0](https://github.com/allenai/allennlp-models/releases/tag/v2.1.0) - 2021-02-24
 
+### Changed
+-  Add option in SnliDatasetReader to support only having "non-entailment" and "entailment" as output labels.
```
Comment (epwalsh) [inline CHANGELOG.md:20]: Can you put this above under the `## Unreleased` section?

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,11 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## Unreleased
 
+### Changed
+-  Add option in SnliDatasetReader to support only having "non-entailment" and "entailment" as output labels.
```
Comment (epwalsh) [inline CHANGELOG.md:11]:
```suggestion
### Changed

-  Add option `combine_input_fields` in `SnliDatasetReader` to support only having "non-entailment" and "entailment" as output labels.
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,11 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
 ### Fixed
 
 - Updated `QaNet` and `NumericallyAugmentedQaNet` models to remove bias for layers that are followed by normalization layers.
 - Updated the model card for `rc-naqanet`.
 - Predictors now work for the vilbert-multitask model
+- Fixing the allennlp issue 4745.
```
Comment (epwalsh) [inline CHANGELOG.md:18]:
```suggestion
```

File: allennlp_models/pair_classification/dataset_readers/snli.py [hunk 5]
```diff
@@ -14,6 +14,19 @@
 logger = logging.getLogger(__name__)
 
 
+def maybe_collapse_label(label: str, collapse: bool):
+    """
+    Helper function that optionally collapses the "contradiction" and "neutral" labels
+    into "non-entailment" and return
+    :param label: one of the NLI task target labels --- "entailment", "contradiction" and "neutral"
+    :param collapse: flag to determine whether to collapse
```
Comment (epwalsh) [inline allennlp_models/pair_classification/dataset_readers/snli.py:22]:
```suggestion
    Helper function that optionally collapses the "contradiction" and "neutral" labels
    into "non-entailment".
```

File: allennlp_models/pair_classification/dataset_readers/snli.py [hunk 6]
```diff
@@ -36,13 +49,19 @@ class SnliReader(DatasetReader):
         If False, represent the premise and the hypothesis as separate fields in the instance.
         If True, tokenize them together using `tokenizer.tokenize_sentence_pair()`
         and provide a single `tokens` field in the instance.
+    combine_input_fields : `bool`, optional
+            (default=`False`)
```
Comment (epwalsh) [inline allennlp_models/pair_classification/dataset_readers/snli.py:53]:
```suggestion
    combine_input_fields : `bool`, optional (default=`False`)
```

File: allennlp_models/pair_classification/dataset_readers/snli.py [hunk 7]
```diff
@@ -36,13 +49,19 @@ class SnliReader(DatasetReader):
...
+        If True, collapse "neutral" and "contradiction" labels into "non-entailment";
+        and left "entailment" unchanged.
```
Comment (epwalsh) [inline allennlp_models/pair_classification/dataset_readers/snli.py:55]:
```suggestion
        If `True`, the "neutral" and "contradiction" labels will be collapsed into "non-entailment";
        "entailment" will be left unchanged.
```

File: allennlp_models/pair_classification/dataset_readers/snli.py [hunk 8]
```diff
@@ -36,13 +49,19 @@ class SnliReader(DatasetReader):
...
+        If False, read labels as they are in the files.
```
Comment (epwalsh) [inline allennlp_models/pair_classification/dataset_readers/snli.py:56]:
```suggestion
```

File: allennlp_models/pair_classification/dataset_readers/snli.py [hunk 9]
```diff
@@ -56,6 +75,9 @@ def __init__(
             self._combine_input_fields = combine_input_fields
         else:
             self._combine_input_fields = isinstance(self._tokenizer, PretrainedTransformerTokenizer)
+        self.collapse_labels = (
+            collapse_labels  # whether turn "neutral", "contradiction" into "non-entailment"
+        )
```
Comment (epwalsh) [inline allennlp_models/pair_classification/dataset_readers/snli.py:80]:
```suggestion
        self.collapse_labels = collapse_labels
```

**Review by epwalsh — COMMENTED** [review]
Looks good! I just left some documentation + style suggestions. See comments ⬇️

File: allennlp_models/pair_classification/dataset_readers/snli.py (hunk 9 above)
Comment (leo-liuzy) [reply allennlp_models/pair_classification/dataset_readers/snli.py:80]: This should be caused by black.

**Comment by leo-liuzy** [comment]
Thanks for all the suggestions!

**Review by leo-liuzy — COMMENTED** [review]
I think this is now finalized?

**Review by epwalsh — APPROVED** [review]
✅
